File: frontend/utils/types/food.ts

export interface Food {
    id: number,
    blsCode: string,
    name: string,
    kcal: number | null,
    water: number | null,
    protein: number | null,
    fat: number | null,
    carbohydrates: number | null
    fiber: number | null
}

export interface FoodWithPortion extends Food {
    portions: FoodPortion[],
}

export interface SearchFood {
    id: number,
    name: string
}

export interface CreateFoodWithAmount {
    foodId: number,
    userId: string,
    amount: number,
    portionId: number | null
    meal: Meal
}

export interface TrackedFood {
    id: number
    food: Food
    amount: number,
    meal: { type: string, typeLabel: string }
    portion: FoodPortion | null
}

export interface CreateTrackedFood {
    id?: number | null;
    foodId: number
    amount: number,
    meal: Meal
    portionId: number | null
}

export interface FoodPortion {
    id: number,
    grams: number,
    label: string,
    isDefault: boolean
}

export interface FoodWithLastEntry extends Food {
    portions: FoodPortion[]
    lastEntry?: {
        amount: number,
    }
}

export interface TrackFoodForUser {
    foodId: number,
    userId: string
    amount: number
}

export enum Meal {
    BREAKFAST,
    LUNCH,
    DINNER,
    SNACK
}

File: backend/src/main/java/com/labs_101/backend/entities/food/Food.java
package com.labs_101.backend.entities.food;

import java.time.Instant;
import java.util.List;

import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@NoArgsConstructor
@AllArgsConstructor
@Getter
@Setter
@Entity
@Table(name = "food")
public class Food {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    private String blsCode;
    private String name;
    private Double kcal;
    private Double water;
    private Double protein;
    private Double fat;
    private Double carbohydrates;
    private Double fiber;

    private String barCode;

    @CreationTimestamp
    private Instant createDate;

    @UpdateTimestamp
    private Instant updateDate;

    @OneToMany(mappedBy = "food")
    private List<TrackedFood> foodUsers;

    @OneToMany(mappedBy = "food", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<FoodPortion> portions;

    public Food(Long foodId) {
        this.id = foodId;
    }

    public Food(String name) {
        this.name = name;
    }

    public void addPortion(FoodPortion portion) {
        this.portions.add(portion);
    }
}
